dep addition checklist (#7325)
Azure/azure-sdk: docs/python/dependency_addition_checklist.md

---
title: "Dependency addition checklist"
permalink: dependency_addition_checklist.html
folder: python
---

# Dependency addition checklist

When adding a dependency to a library, it's crucial to ensure that the dependency aligns with the library's goals, technical requirements, and legal constraints.

> Approved dependencies can be found at https://github.com/Azure/azure-sdk/blob/main/docs/python/approved_dependencies.md

Reviewer: add typing-extensions

Reviewer 2: We should also make sure the this is *in addition to*, not *instead of* any other requirements listed elsewhere.

### Review checklist

- DO work with architects to review the intent of how you want to use it. e.g. sync/async, etc.
- DO work with LCA team to review the license compatibility. (aka.ms/cela -> Find my CELA contact)
  - MIT is the only one to auto-approve

> - The library must have stable releases.

Reviewer 3: Library must also support the full band of supported Python versions.

Reviewer 2: We should always support both sync and async. So if the dependency only supports sync (and does IO), they have to have an answer for async (or the other way around)

> - The library must be pure Python or it provides wheels for all platforms and environments our SDK supports.

Reviewer 2: ```suggestion
- The library must be pure Python, or provide wheels for all platforms and environments our SDK supports.
```

> - The library must be in active development and be actively maintained, with regular updates and patches.

Reviewer 2: I would prefer a more authoritative language (i.e. not use "ideally").

> - The library must be in active development and be actively maintained, with regular updates and patches.

Reviewer 3: Good point - updated my comment-patch to drop "ideally" :)

- NO dependency conflicts. Users must be able to use all our SDKs in the same environment. (We have some coverage in CI)

> - DO understand the liability that you need to update your SDK in a timely manner if your SDK is broken due to a new release of the dependency.

Reviewer 2: Expectation on "timely" is closer to days than months.

> - DO data protection check: if the dependency handles data, ensure it does so in a manner that complies with data protection laws applicable to your users.

Reviewer 3: ```suggestion
- DO data protection check: if the dependency handles data, ensure it does so in a manner that complies with global data protection laws applicable to your users.
```

---
title: "Dependency addition checklist"
permalink: dependency_addition_checklist.html
folder: python
---

# Dependency addition checklist

When adding a dependency to a library, it's crucial to ensure that the dependency aligns with the library's goals, technical requirements, and legal constraints.

> - You can directly use the library if it's on [approved dependencies list](https://github.com/Azure/azure-sdk/blob/main/docs/python/approved_dependencies.md).

Reviewer 3: ```suggestion
- You can simply use the library directly if it's on the [approved dependencies list](https://github.com/Azure/azure-sdk/blob/main/docs/python/approved_dependencies.md).
```

> - You can bypass the checklist for libraries on the [conditionally approved dependencies list](https://github.com/Azure/azure-sdk/blob/main/docs/python/conditionally_approved_dependencies.md), but ensure to consult with the architects about your intended use of the library.

Reviewer 3: ```suggestion
- For libraries on the [conditionally approved dependencies list](https://github.com/Azure/azure-sdk/blob/main/docs/python/conditionally_approved_dependencies.md), the checklist can be ignored, but be sure to consult with the architects about your intended use of the library.
```

- The aiohttp library does not require approval for HTTP usage; however, consult with the architects before using it for websocket purposes.
- If the library does not meet all of these requirements below, you could discuss with the architects about using an optional dependency to enable certain opt-in scenarios.

### Review checklist

> - DO work with architects to review the intent of how you want to use it. e.g. sync/async, etc.

Reviewer 3: ```suggestion
- DO work with architects to review how you intend to use it, e.g. sync/async, etc.
```
